react/features/transcription/reducer.js

```diff
@@ -0,0 +1,122 @@
+import {
+    ENDPOINT_MESSAGE_RECEIVED,
+    ADD_TRANSCRIPT_MESSAGE,
+    UPDATE_TRANSCRIPT_MESSAGE,
+    REMOVE_TRANSCRIPT_MESSAGE
+} from './actionTypes';
+import { getUpdatedTranscriptionParagraphs } from './functions';
+import { ReducerRegistry } from '../base/redux';
+
+/**
+ * Default State for 'features/transcription' feature
+ */
+const defaultState = {
+    transcriptMessages: {},
+    transcriptionSubtitles: []
+};
+
+/**
+ * Listen for actions for the transcription feature to be used by the actions
+ * to update the rendered transcription subtitles.
+ */
+ReducerRegistry.register('features/transcription', (
+        state = defaultState, action) => {
+
+    switch (action.type) {
+    case ENDPOINT_MESSAGE_RECEIVED:
+        return _endpointMessageReceived(state);
+
+    case ADD_TRANSCRIPT_MESSAGE:
+        return _addTranscriptMessage(state, action);
+
+    case UPDATE_TRANSCRIPT_MESSAGE:
+        return _updateTranscriptMessage(state, action);
+
+    case REMOVE_TRANSCRIPT_MESSAGE:
+        return _removeTranscriptMessage(state, action);
+    }
+
+    return state;
+});
+
+/**
+ * Reduces a specific Redux action ENDPOINT_MESSAGE_RECEIVED of the feature
+ * transcription.
+ *
+ * @param {Object} state - The Redux state of the feature transcription.
+ * @returns {Object} The new state of the feature transcription after the
+ * reduction of the specified action.
+ */
+function _endpointMessageReceived(state) {
+
+    return {
+        ...state,
+        transcriptionSubtitles:
+            getUpdatedTranscriptionParagraphs(state.transcriptMessages)
```

Comment:
As yanas had mentioned to remove too many calculations from the Component's `render()` method, I moved it to precalculate the store the array of subtitles paragraphs and just passed it as a `prop`.
This way rendering will be faster as it will have less javascript code to process.

Reply:
I can't agree with this optimization as I don't see how it is an optimization if on every update of messages this gets run again anyway, which is essentially a re-render to me. I may not be seeing some skips of calculation that this optimization adds. What I do see is storing extra truth to the store, which is essentially global app state, that won't be used by any other components except the subtitles. If you want to prevent extra re-renders you might want to look into shouldComponentUpdate or using a PureComponent.

I'd ask for additional opinions on this one.